torch_np/_dtypes_impl.py

```diff
@@ -49,3 +49,80 @@ def result_type_impl(*tensors):
         dtyp = _cd._result_type_dict[dtyp][curr.dtype]
 
     return dtyp
+
+
+# ### NEP 50 helpers ###
+
+SCALAR_TYPES = (int, bool, float, complex)
+
+
+def _dtype_for_scalar(py_type):
+    return {
+        bool: torch.bool,
+        int: torch.int64,
+        float: torch.float64,
+        complex: torch.complex128,
+    }[py_type]
+
+
+categories = [
+    (torch.bool,),
+    (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64),
+    (torch.float16, torch.float32, torch.float64),
+    (torch.complex64, torch.complex128),
+]
+
+
+def category(dtyp):
+    for j, cat in enumerate(categories):
+        if dtyp in cat:
+            return j
+    raise ValueError(f"unknown dtype {dtyp}")
+
+
+dtype_for_cat = {0: torch.bool, 1: torch.int64, 2: torch.float64, 3: torch.complex128}
+
+
+def nep50_to_tensors(x1, x2):
+    """If either of inputs is a python scalar, type-promote with NEP 50.
+
+    NB: NEP 50 mandates RuntimeWarnings on some overflows. We do not emit them:
+    we either raise OverflowError or just do the computation.
+    """
+
+    x1_type, x2_type = type(x1), type(x2)
+    x1_is_weak = x1_type in SCALAR_TYPES
+    x2_is_weak = x2_type in SCALAR_TYPES
+    if x1_is_weak and x2_is_weak:
+        # two scalars: promote
+        x1 = torch.as_tensor(x1, dtype=_dtype_for_scalar(x1_type))
+        x2 = torch.as_tensor(x2, dtype=_dtype_for_scalar(x2_type))
+        return x1, x2
+    elif not (x1_is_weak or x2_is_weak):
+        # two tensors: nothing to do here
+        return x1, x2
+    else:
+        # scalar <op> scalar: NEP 50
+        weak, not_weak = (x1, x2) if x1_is_weak else (x2, x1)
+
+        # find the dtype for the weak's type
+        weak_dtype = _dtype_for_scalar(type(weak))
+
+        cat_weak = category(weak_dtype)
+        cat_not_weak = category(not_weak.dtype)
+
+        dt = not_weak.dtype if cat_weak <= cat_not_weak else dtype_for_cat[cat_weak]
+
+        # special-case complex + float32
+        if weak_dtype.is_complex and not_weak.dtype == torch.float32:
+            dt = torch.complex64
```

Comment:
Well, i suppose you don't have long-double, so there are only two complex dtypes and this thus covers all N=1 cases :).

It would be sensible to special case this to weak_complex + float[prec] -> complex[prec]  (prec*2 depending on your notation).  But maybe this is pragmatic (maybe a commont if you worry)?

Reply:
Exactly. Not having longdouble makes things easier :-). Since it's indeed the only case, I'd just keep the special case and move on.